Orange/widgets/unsupervised/owkmeans.py

```diff
@@ -172,114 +229,185 @@ def adjustSize(self):
-            if all(isinstance(score, str)
-                   for score in self.clusterings.values()):
-                # Tooltip shows just the last error
-                # pylint: disable=undefined-loop-variable
-                self.Error.failed(self.clusterings[k])
-                self.mainArea.hide()
-        finally:
-            self.controlArea.setDisabled(False)
-        return True
+        if not self.enough_data_instances(self.k_from):
+            self.Error.not_enough_data(len(self.data), self.k_from)
+            return
+
+        if not self.enough_data_instances(self.k_to):
+            self.Warning.not_enough_data(len(self.data), self.k_to)
+            return
+
+        needed_ks = [k for k in range(self.k_from, self.k_to + 1)
+                     if k not in self.clusterings]
+
+        if needed_ks:
+            self.__launch_tasks(needed_ks)
+        else:
+            # If we don't need to recompute anything, just set the results to
+            # what they were before
+            self.update_results()
 
     def cluster(self):
-        if self.k in self.clusterings or \
-                not self.check_data_size(self.k, self.Error):
+        # Check if the k already has a computed clustering
+        if self.k in self.clusterings:
+            return
+
+        # Check if there is enough data
+        if not self.enough_data_instances(self.k):
+            self.Error.not_enough_data(len(self.data), self.k)
             return
-        try:
-            self.controlArea.setDisabled(True)
-            if not self._compute_clustering(self.k):
-                self.Error.failed(self.clusterings[self.k])
-        finally:
-            self.controlArea.setDisabled(False)
-
-    def apply(self):
+
+        self.__launch_tasks([self.k])
+
+    def commit(self):
+        self.cancel()
         self.clear_messages()
-        if self.data is not None:
-            if self.optimize_k and self.run_optimization():
-                self.mainArea.show()
-                self.update_results()
-            else:
-                self.cluster()
-                self.mainArea.hide()
+
+        # Some time may pass before the new scores are computed, so clear the
+        # old scores to avoid potential confusion. Hiding the mainArea could
+        # cause flickering when the clusters are computed quickly, so this is
+        # the better alternative
+        self.table_model.clear_scores()
+        self.mainArea.setVisible(self.optimize_k and self.data is not None)
+
+        if self.data is None:
+            self.send_data()
+            return
+
+        if self.optimize_k:
+            self.run_optimization()
         else:
-            self.mainArea.hide()
+            self.cluster()
+
         QTimer.singleShot(100, self.adjustSize)
-        self.send_data()
 
-    def invalidate(self, force_apply=False):
+    def invalidate(self, force=False):
         self.Error.clear()
         self.Warning.clear()
         self.clusterings = {}
```

Comment:
`cancel` pending tasks here regardless of `auto_commit` state. If any are still in flight and are not cancelled then `self.clusterings` will be updated in `__clustering_complete` with now invalid results.